### PR TITLE
feat(ui): Add `StateLayer`

### PR DESCRIPTION
**Pull request by vuluongj20**
Add `StateLayer`, a visual component for indicating hover/press state. It's a replacement for our current approach of changing the background color on hover/press. 

There are 2 main advantages to using `StateLayer`:
* **It supports both hover and press states.** Currently we only have colors for hover states. We could add more for press states, but that would introduce more complexity to our color system than I'd prefer. `StateLayer` doesn't add any new color – it just uses the current text color.
  Screen Shot 2022-11-16 at 3 14 50 PM
* **It works on most backgrounds.** Our current hover colors work fine on many backgrounds but look dimmed on some other. `StateLayer` avoids that with its opacity-based approach:
  Screen Shot 2022-11-16 at 3 08 29 PM
  Screen Shot 2022-11-16 at 3 08 50 PM
  Screen Shot 2022-11-16 at 3 20 15 PM
  Screen Shot 2022-11-16 at 3 19 44 PM

### CODE REVIEW

File: static/app/components/stateLayer.tsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  border-color: currentcolor;
+  pointer-events: none;
+
+  opacity: 0;
+
+  ${p =>
+    defined(p.isHovered)
+      ? `
+        *:not(.focus-visible) > & {
+          opacity: ${p.isHovered ? (p.higherOpacity ? 0.085 : 0.06) : 0};
```
Comment (malwilley) [inline static/app/components/stateLayer.tsx:38]: Should these numbers be in the theme? Not sure myself, just putting the question out there

File: static/app/components/stateLayer.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import {defined} from 'sentry/utils';
+
+interface StateLayerProps extends React.HTMLAttributes<HTMLSpanElement> {
+  color?: string;
+  higherOpacity?: boolean;
+  isHovered?: boolean;
+  isPressed?: boolean;
+}
+
+const StateLayer = styled(
```
Comment (malwilley) [inline static/app/components/stateLayer.tsx:13]: I think this name could be a bit more descriptive... `InteractableStateLayer` maybe?

File: static/app/components/stateLayer.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import isPropValid from '@emotion/is-prop-valid';
+import styled from '@emotion/styled';
+
+import {defined} from 'sentry/utils';
+
+interface StateLayerProps extends React.HTMLAttributes<HTMLSpanElement> {
+  color?: string;
+  higherOpacity?: boolean;
```
Comment (malwilley) [inline static/app/components/stateLayer.tsx:8]: Not a huge fan of needing to pass this, but I see why it's necessary and can't think of a way around it. Seems like an easy way for things to become inconsistent 😕

**Review by malwilley — APPROVED** [review]
Love the addition of active states, have been wanting those!

And I like the idea of overlaying the text color, that's a clever way to make this a lot simpler.

File: static/app/components/stateLayer.tsx (hunk 1 above)
Comment (vuluongj20) [reply static/app/components/stateLayer.tsx]: Currently these numbers are only used by `StateLayer` and I don't see them being useful elsewhere, so I think it's best that they're here. In the future we can move them to theme if there's another component that needs them.

File: static/app/components/stateLayer.tsx (hunk 2 above)
Comment (vuluongj20) [reply static/app/components/stateLayer.tsx]: Hmm good point actually. Named it `InteractionStateLayer`, as "Interaction State" is a commonly used term in design systems.

File: static/app/components/stateLayer.tsx (hunk 3 above)
Comment (vuluongj20) [reply static/app/components/stateLayer.tsx]: Yeah I agree, lots of room for inconsistencies. One thing that I considered but decided against is to automate this by using the `color` package to check the contrast level between the current text color and the background color. If the contrast is below a certain amount then we enforce higher opacity. But that would involve a bunch of style queries and seems too bloated for its purpose.

**Pull request merged**